File: solutions/py/11724.py
import sys

N, M = map(int, sys.stdin.readline().split())

graph = [ [] for _ in range(N+1) ]
visited = [False]*(N+1)
visited[0] = True

for _ in range(M):
    u, v = map(int, sys.stdin.readline().split())
    graph[u].append(v)
    graph[v].append(u)

result = 0
current = 1
while not all(visited):
    result += 1
    while visited[current]:
        current += 1

    stack = [current]

    while stack:
        vertex = stack.pop()
        visited[vertex] = True
        for v in graph[vertex]:
            if not visited[v]:
                stack.append(v)

print(result)

File: solutions/rb/11724.rb
N, M = gets.chomp.split.map(&:to_i)

graph = Array.new(N+1) { [] }
visited = Array.new(N+1, false)
visited[0] = true

(1..M).each do
  u, v = gets.chomp.split.map(&:to_i)
  graph[u].push(v)
  graph[v].push(u)
end

count = 0
current = 1
until visited.all?
  count += 1
  while visited[current]
    current += 1
  end

  stack = [current]

  until stack.empty?
    vertex = stack.pop
    visited[vertex] = true
    graph[vertex].each do |v|
      if !visited[v]
        stack.push(v)
      end
    end
  end
end

puts count
